### ISSUE TITLE
Extreme performance of the waste when CONCURRENT_ITEMS is large

### ISSUE THREAD

**Issue by MagicalBomb** [scrapy/scrapy#5182]
### Description

There is  Extreme performance of the waste When CONCURRENT_ITEMS set to a ge large number, such as 9999.

Some days ago, I wrote a spider with CONCURRENT_ITEMS=9999, and run it. I found my spider use 100% cpu and crawl speed is very very slow （40 pages/mins）, normally it would be 400+pages/min.

### Steps to Reproduce

Here is code , you can run directly. I would see cpu up to 100%. Wait 1 minute, you will see 
[scrapy.extensions.logstats] INFO: Crawled 27 pages (at 27 pages/min)

```

import scrapy


class TestSpider(scrapy.Spider):
    name = 'test'

    custom_settings = {
        "LOG_LEVEL": "INFO",
        "CONCURRENT_ITEMS": 99999,
    }

    def start_requests(self):
        for _ in range(3000):
            yield scrapy.Request(
                url="http://httpbin.org/get",
                dont_filter=True,
            )

    def parse(self, response):
        pass

```

**Comment by Gallaecio** [comment]
I suspect that might be to be expected.

I believe the root code for this is https://github.com/scrapy/scrapy/blob/7e23677b52b659b11471a63f3be9905a0bbaf995/scrapy/utils/defer.py#L72 (`count` comes from `CONCURRENT_ITEMS`). Quoting the first paragraph of the blog post linked from there as the source of this implementation:

> Concurrency can be a great way to speed things up, but what happens when you have too much concurrency? Overloading a system or a network can be detrimental to performance.

So maybe the best way forward here is to treat this as a documentation issue, and document how increasing this number may be counter-productive.

I also wonder if it would make sense to count the number of elements in `iterable` (assuming they are not infinite) and limit `count` to the lesser value, i.e. `count = min(count, iterable_length)`. Maybe we could even implement this while supporting an arbitrarily large iterable, by peaking items from iterable up to `count`.

**Comment by Gallaecio** [comment]
Sounds interesting. I would not mind allowing a special value for the `CONCURRENT_ITEMS` option, e.g. `auto`, that applies this logic. If it works well, and it sounds like it should, we might want to consider making it the default value at some point. It could also long the final value chosen, so that if the ramp-up time is a problem, users can replace `auto` with an actual value based on the logged values from prior executions.

**Comment by edorado93** [comment]
Yep, that sounds good to me. An `auto` configuration would be cool to have (and as default in the future!). And yeah, logging also is a great idea to avoid the same computations by scrappy and so that users can provide the optimal value themselves. 

I'd love to work on this :) Is it ok if I assign this to myself?

**Comment by Gallaecio** [comment]
We don’t tend to assign issues except for core maintainers, and even then rarely. Your comment above should be enough to let people know someone is working on this, though.

I’m really looking forward to the results. I’m also hoping we can do some benchmarks once implemented, to see after how many items it becomes worth it to use `auto` vs the current default value.
